STM32  nor flash controller gets address and size

FRASTM opened the pull request:
the stm32  qspi and ospi flash drivers can retrieve the NOR flash address and size directly from the device tree 
of their `compatible = "st,stm32-qspi-nor";`  or `compatible = "st,stm32-ospi-nor";`
This is done with the <reg> property of the qspi-nor-flash or ospi-nor-flash node as follows:

```
 mx25lm51245: ospi-nor-flash@70000000 {
                 compatible = "st,stm32-ospi-nor";
                reg = <0x70000000 DT_SIZE_M(64)>; /* 512 Mbits */
```

That will make the nor-flash controller driver using the address and size property and will prepare the flash controller for the memory mapped mode (XIP).
That will avoid confusion around the <size> expressed in bits or in byte : here is the reg second param a size in Bytes.

FRASTM committed:
drivers: flash: stm32 ospi drivers gets address and size from DTS

Address and size are given by the DTS register property
of the ospi nor : to be used by the ospi driver.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32 disco kit has ospi nor flash node

Define the reg and size property for the stm32u585 and stm32l562
disco kit.
Refer to the dts/bindings/flash_controller/st,stm32-ospi-nor.yaml.
The stm32l562 reads the sfdp table from the flash itself.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: bindings: flash controller stm32 qspi nor flash reg property

Address and size are given by the DTS register property
of the qspi nor. The size Property becomes useless.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
boards: arm: stm32 boards has qspi nor flash node

Define the reg and size property for the stm32 boards with qspi inside
Refer to the dts/bindings/flash_controller/st,stm32-ospi-nor.yaml.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
dts: bindings: flash controller stm32 ospi nor flash reg property

Address and size are given by the DTS register property
of the ospi nor. The size Property becomes useless

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: flash: stm32 qspi drivers gets address and size from DTS

Address and size are given by the DTS register property
of the qspi nor : to be used by the qspi driver.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango reviewed (commented):
Please also update migration guide `doc/releases/migration-guide-3.6.rst`

FRASTM committed:
doc: releases:; migration guide updated

Update the migration guide for release v3.6 to detail the change about
the bindings of `st,stm32-ospi-nor` and `st,stm32-qspi-nor` compatible

Signed-off-by: Francois Ramu <francois.ramu@st.com>

Laczen reviewed (dismissed):
LGTM, it is nice to see the size removed.

erwango commented on drivers/flash/flash_stm32_ospi.c:
```diff
@@ -2300,7 +2307,7 @@ static const struct flash_stm32_ospi_config flash_stm32_ospi_cfg = {
 #endif
 	.irq_config = flash_stm32_ospi_irq_config_func,
-	.flash_size = DT_INST_PROP(0, size) / 8U,
+	.flash_size = DT_REG_ADDR_BY_IDX(DT_INST(0, st_stm32_ospi_nor), 1),
```
What about using `DT_INST_REG_ADDR_BY_IDX` directly ?

erwango reviewed (changes requested):
Otherwise LGTM

FRASTM commented:
same with DT_INST_REG_ADDR(0) and DT_INST_REG_ADDR_BY_IDX(0, 1)  macro

erwango reviewed (approved):
Thanks

erwango commented:
CI issue under fix in https://github.com/zephyrproject-rtos/zephyr/pull/68285

fabiobaltieri commented:
```
/__w/zephyr/zephyr/samples/sensor/lsm6dsl/src/main.c:150:59: error: implicit conversion from 'float' to 'double' when passing argument to function [-Werror=double-promotion]
  150 |                                                           out_ev(&accel_x_out),
      |                                                           ^~~~~~~~~~~~~~~~~~~~
```

CI is sad

erwango commented:
> CI is sad

CI issue under fix in https://github.com/zephyrproject-rtos/zephyr/pull/68285
